lib/nostrum/api/ratelimiter.ex

```diff
@@ -9,188 +9,374 @@ defmodule Nostrum.Api.Ratelimiter do
+      retry: 3,
+      tls_handshake_timeout: :timer.seconds(5),
+      tls_opts: Constants.gun_tls_opts()
+    }
+
+    {:ok, conn_pid} = :gun.open(domain, 443, open_opts)
+    {:keep_state, %{data | conn: conn_pid}}
   end
 
-  defp get_retry_time(route, method) do
-    route
-    |> get_endpoint(method)
-    |> RatelimitBucket.timeout_for()
+  def connecting(:info, {:gun_up, conn_pid, :http2}, %{conn: conn_pid} = data) do
+    {:next_state, :connected, data}
   end
 
-  @doc """
-  Queue the given request.
+  def connecting({:call, _from}, _request, _data) do
+    {:keep_state_and_data, :postpone}
+  end
 
-  If the ratelimiter tells us to sit it out and we have more than `0` attempts
-  remaining, we sleep out the given retry time and ask it to queue again
-  afterwards.
-  """
-  def queue(request, attempts_remaining \\ @default_attempts_to_requeue) do
-    case GenServer.call(@registered_name, {:queue, request}) do
-      {:error, {:retry_after, time}} when attempts_remaining > 0 ->
-        truncated = :erlang.ceil(time)
+  def connecting(:state_timeout, :connect_timeout, _data) do
+    {:stop, :connect_timeout}
+  end
 
-        attempt = @default_attempts_to_requeue - attempts_remaining + 1
+  def connected({:call, from}, {:queue, request}, %{outstanding: outstanding} = data) do
+    bucket = get_endpoint(request.route, request.method)
+    entry = {request, from}
 
-        Logger.info(
-          "RATELIMITER: Waiting #{truncated}ms to process request with route #{request.route} (try #{attempt} / #{@default_attempts_to_requeue})"
-        )
+    case :maps.get(bucket, outstanding, nil) do
+      nil ->
+        # The case of no queue needs to be handled specially because if we already have a queue
+        # then the requests are gonna start each other as soon as they finish.
+        queue = :queue.in(entry, :queue.new())
 
-        Process.sleep(truncated)
-        queue(request, attempts_remaining - 1)
+        {:keep_state, %{data | outstanding: :maps.put(bucket, queue, outstanding)},
+         [{{:timeout, bucket}, 0, :next}]}
 
-      # We've had enough. Bail.
-      {:error, {:retry_after, _time}} when attempts_remaining == 0 ->
-        {:error, :max_retry_attempts_exceeded}
+      queue ->
+        {:keep_state, %{data | outstanding: %{outstanding | bucket => :queue.in(entry, queue)}}}
+    end
+  end
 
-      result ->
-        result
+  def connected(
+        {:timeout, bucket},
+        :next,
+        %{conn: conn, running: running, outstanding: outstanding} = data
+      )
+      when is_map_key(outstanding, bucket) do
+    case :queue.out(:maps.get(bucket, outstanding)) do
+      {{:value, {request, from}}, queue} ->
+        stream =
+          Base.request(
+            conn,
+            request.method,
+            request.route,
+            request.body,
+            request.headers,
+            request.params
+          )
+
+        running_with_this = :maps.put(stream, {bucket, request, from}, running)
```

Comment:
Could you convert the calls to `:maps` into calls to `Map`? Just to avoid confusion since they usually have different argument ordering?

Reply:
This is also a good idea. I have also updated this.